using-timescaledb/custom-jobs.md

```diff
@@ -0,0 +1,221 @@
+END
+$$;
+```
+
+Register job to run daily dropping chunks on all hypertables that are older
+than 12 months.
+
+```sql
+SELECT add_job('generic_retention','1d', config => '{"lag":"12 month"}');
+```
+
+### Tiered storage [](tiered-storage)
+
+Custom job that moves chunks older than a certain time to a different tablespace.
+
+```sql
+CREATE OR REPLACE PROCEDURE move_chunks (job_id int, config jsonb)
+LANGUAGE PLPGSQL
+AS $$
+DECLARE
+  ht REGCLASS;
+  lag interval;
+  destination name;
+  chunk REGCLASS;
+  tmp_name name;
+BEGIN
+  SELECT jsonb_object_field_text (config, 'hypertable')::regclass INTO STRICT ht;
+  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
+  SELECT jsonb_object_field_text (config, 'tablespace') INTO STRICT destination;
+
+  IF ht IS NULL OR lag IS NULL OR destination IS NULL THEN
+    RAISE EXCEPTION 'Config must have hypertable, lag and destination';
+  END IF;
+
+  FOR chunk IN
+  SELECT show.oid
+  FROM show_chunks(ht, older_than => lag)
+  SHOW (oid)
+    INNER JOIN pg_class pgc ON pgc.oid = show.oid
+    INNER JOIN pg_tablespace pgts ON pgts.oid = pgc.reltablespace
+  WHERE pgts.spcname != destination;
+  LOOP
+    RAISE NOTICE 'Moving chunk: %', chunk::text;
+    EXECUTE format('ALTER TABLE %s SET TABLESPACE %I;', chunk, destination);
+  END LOOP;
+END
+$$;
+```
+
+Register job to run daily moving chunks older than 12 months on hypertable
+metrics to tablespace old_chunks.
+
+```sql
+SELECT add_job('move_chunks','1d', config => '{"hypertable":"metrics","lag":"12 month","tablespace":"old_chunks"}');
+```
+
+### Downsample and compress [](downsample-compress)
+
+Custom job that downsamples and compresses chunks on hypertable metrics older
+than a certain age.
+
+```sql
+CREATE OR REPLACE PROCEDURE downsample_compress (job_id int, config jsonb)
+LANGUAGE PLPGSQL
+AS $$
+DECLARE
+  lag interval;
+  chunk REGCLASS;
+  tmp_name name;
+BEGIN
+  SELECT jsonb_object_field_text (config, 'lag')::interval INTO STRICT lag;
+
+  IF lag IS NULL THEN
+    RAISE EXCEPTION 'Config must have lag';
+  END IF;
+
+  FOR chunk IN
+    SELECT show.oid
+    FROM show_chunks('metrics', older_than => lag) SHOW (oid)
+      INNER JOIN pg_class pgc ON pgc.oid = show.oid
+      INNER JOIN pg_namespace pgns ON pgc.relnamespace = pgns.oid
+      INNER JOIN timescaledb_information.chunks chunk ON chunk.chunk_name = pgc.relname
+        AND chunk.chunk_schema = pgns.nspname
+    WHERE chunk.is_compressed::bool = FALSE
+  LOOP
+    RAISE NOTICE 'Processing chunk: %', chunk::text;
+
+    -- build name for temp table
+    SELECT '_tmp' || relname
+    FROM pg_class
+    WHERE oid = chunk INTO STRICT tmp_name;
+
+    -- copy downsampled chunk data into temp table
+    EXECUTE format($sql$ CREATE UNLOGGED TABLE %I AS
+      SELECT time_bucket('1h', time), device_id, avg(value) FROM %s GROUP BY 1, 2;
+    $sql$, tmp_name, chunk);
+
+    -- clear original chunk
+    EXECUTE format('TRUNCATE %s;', chunk);
```

Comment:
Curious for my own education -- why TRUNCATE the chunk rather than `DROP TABLE` it ?  Any difference?  (If identical, maybe use the latter as raises fewer potential questions?)

Reply:
TRUNCATE keeps the chunk itself intact, since we do the insert into the chunk we can skip the chunk routing that way. It could be rewritten to use DROP TABLE but that would recreate the chunk